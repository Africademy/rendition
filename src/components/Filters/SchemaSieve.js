import * as assign from 'lodash/assign'
import * as cloneDeep from 'lodash/cloneDeep'
import * as isArray from 'lodash/isArray'
import * as every from 'lodash/every'
import * as some from 'lodash/some'
import * as isPlainObject from 'lodash/isPlainObject'
import * as pickBy from 'lodash/pickBy'
import * as forEach from 'lodash/forEach'
import * as map from 'lodash/map'
import filterTests from '../PineTypes'

export const SIMPLE_SEARCH_NAME = 'Full text search'

class SchemaSieve {
  constructor (tests = {}) {
    this.tests = assign(cloneDeep(filterTests), tests)

    this.SIMPLE_SEARCH_NAME = SIMPLE_SEARCH_NAME
  }

  test (item, input) {
    return isArray(input)
      ? every(input, i => this.baseTest(item, i))
      : this.baseTest(item, input)
  }

  baseTest (item, input) {
    const { type, name } = input

    // A simple search is not strictly a "type" and searches on all fields,
    // so we handle that seperately here
    if (name === SIMPLE_SEARCH_NAME) {
      const { value } = input
      return some(
        item,
        target =>
          target &&
          String(target)
            .toLowerCase()
            .includes(value.toLowerCase())
      )
    }

    if (type in filterTests) {
      const { operator, value } = input
      const target = item[name]

      if (operator in filterTests[type].rules) {
        const result = filterTests[type].rules[operator]
        return isPlainObject(result)
          ? result.test(target, value)
          : result(target, value)
      }
    }

    // If the rule contains an unrecognised type or operator, then play it safe
    // and don't filter the item
    return true
  }

  filter (collection, input) {
    if (isArray(collection)) {
      return this.filterArray(collection, input)
    }
    if (isPlainObject(collection)) {
      return this.filterObject(collection, input)
    }

    throw new Error('collection argument must be either object or array.')
  }

  filterArray (collection, input) {
    return collection.filter(item => this.test(item, input))
  }

  filterObject (collection, input) {
    return pickBy(collection, value => this.test(value, input))
  }

  getOperators (type, schemaEntry) {
    if (type in filterTests) {
      // If the rule has a 'getLabel' method, use that to construct the return object
      return map(filterTests[type].rules, (value, key) => {
        if (isPlainObject(value) && value.getLabel) {
          const label = value.getLabel(schemaEntry)
          if (label) {
            return {
              label,
              value: key
            }
          }
        }
        return {
          label: key,
          value: key
        }
      })
    }
    return []
  }

  makeFilterInputs (schema) {
    const inputs = {}

    forEach(schema, (value, key) => {
      inputs[key] = {
        type: value.type,
        name: key,
        label: value.label,
        availableOperators: this.getOperators(value.type, schema[key]),
        operator: null,
        value: null
      }
    })

    return inputs
  }

  validate (type, value) {
    // If the type is unknown just return true
    if (!(type in filterTests)) {
      return true
    }

    return filterTests[type].validate(value)
  }

  getTypes () {
    return Object.keys(filterTests)
  }
}

export default tests => new SchemaSieve(tests)
