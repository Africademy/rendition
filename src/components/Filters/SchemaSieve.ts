import assign = require('lodash/assign');
import cloneDeep = require('lodash/cloneDeep');
import every = require('lodash/every');
import forEach = require('lodash/forEach');
import includes = require('lodash/includes');
import isArray = require('lodash/isArray');
import isPlainObject = require('lodash/isPlainObject');
import map = require('lodash/map');
import omit = require('lodash/omit');
import pickBy = require('lodash/pickBy');
import some = require('lodash/some');
import {
	AdvancedPineOperatorTest,
	FilterModel,
	FilterRule,
	PineTypeModule,
	Schema,
	SchemaEntry,
	SchemaSieveClass,
} from 'rendition';
import filterTests from '../PineTypes';

type PineTypeModuleStruct = { [key: string]: PineTypeModule };

export const SIMPLE_SEARCH_NAME = 'Full text search';

class SchemaSieve implements SchemaSieveClass {
	public SIMPLE_SEARCH_NAME = SIMPLE_SEARCH_NAME;
	public tests: PineTypeModuleStruct;

	constructor(tests: PineTypeModuleStruct = {}) {
		this.tests = assign(cloneDeep(filterTests), tests);
	}

	test(item: any, input: FilterRule | FilterRule[]) {
		return isArray(input)
			? every(input, i => this.baseTest(item, i))
			: this.baseTest(item, input);
	}

	baseTest(item: any, input: FilterRule): boolean {
		// If this is a compound rule, evaluate the rules together using OR logic
		if (input.extra && input.extra.or) {
			return (
				this.baseTest(item, omit(input, 'extra') as FilterRule) ||
				some(input.extra.or, rule => this.baseTest(item, rule as FilterRule))
			);
		}

		const { type, name } = input;

		// A simple search is not strictly a "type" and searches on all fields,
		// so we handle that seperately here
		if (name === this.SIMPLE_SEARCH_NAME) {
			const { value } = input;
			return some(
				item,
				target =>
					target && includes(String(target).toLowerCase(), value.toLowerCase()),
			);
		}

		if (type in filterTests) {
			const { operator, value } = input;
			const target = item[name];

			if (operator! in filterTests[type].rules) {
				const result = filterTests[type].rules[operator!];
				return isPlainObject(result)
					? result.test(target, value)
					: result(target, value);
			}
		}

		// If the rule contains an unrecognised type or operator, then play it safe
		// and don't filter the item
		return true;
	}

	filter<T>(
		collection: T[],
		input: FilterRule | FilterRule[],
	): T[] | Partial<T> {
		if (isArray(collection)) {
			return this.filterArray(collection, input);
		}
		if (isPlainObject(collection)) {
			return this.filterObject(collection, input) as Partial<T>;
		}

		throw new Error('collection argument must be either object or array.');
	}

	filterArray(collection: any[], input: FilterRule | FilterRule[]) {
		return collection.filter(item => this.test(item, input));
	}

	filterObject(collection: any, input: FilterRule | FilterRule[]) {
		return pickBy(collection, value => this.test(value, input));
	}

	getOperators(type: string, schemaEntry: SchemaEntry) {
		if (type in filterTests) {
			// If the rule has a 'getLabel' method, use that to construct the return object
			return map((filterTests[type] as PineTypeModule).rules, (value, key) => {
				if (!!(<AdvancedPineOperatorTest>value).getLabel) {
					const label = (<AdvancedPineOperatorTest>value).getLabel(schemaEntry);
					if (label) {
						return {
							label,
							value: key,
						};
					}
				}
				return {
					label: key,
					value: key,
				};
			});
		}
		return [];
	}

	makeFilterInputs(schema: Schema) {
		const inputs: {
			[key: string]: FilterModel;
		} = {};

		forEach(schema, (value, key) => {
			inputs[key] = {
				type: value.type,
				name: key,
				label: value.label,
				availableOperators: this.getOperators(value.type, schema[key]),
				operator: null,
				value: null,
			};
		});

		return inputs;
	}

	validate(type: string, value: any) {
		// If the type is unknown just return true
		if (!(type in filterTests)) {
			return true;
		}

		return filterTests[type].validate(value);
	}

	getTypes() {
		return Object.keys(filterTests);
	}
}

export default (tests?: PineTypeModuleStruct) => new SchemaSieve(tests);
