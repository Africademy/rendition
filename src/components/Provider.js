import * as React from 'react'
import styled, { ThemeProvider } from 'styled-components'
import defaultTheme from '../theme'

const Base = styled.div`
  font-family: ${props => props.theme.font};
`

const Provider = ({ theme, ...props }) => {
  return (
    <ThemeProvider theme={theme || defaultTheme}>
      <Base {...props} />
    </ThemeProvider>
  )
}

export default Provider
