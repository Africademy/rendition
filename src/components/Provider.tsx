import * as React from 'react';
import styled, { StyledFunction, ThemeProvider } from 'styled-components';
import { DefaultProps, Theme } from '../common-types';
import defaultTheme from '../theme';

export interface ThemedBaseProps extends DefaultProps {
	theme?: Theme;
}

const Base = (styled.div as StyledFunction<DefaultProps>)`
  font-family: ${props => props.theme.font};
`;

const Provider = ({ theme, ...props }: ThemedBaseProps) => {
	return (
		<ThemeProvider theme={theme || defaultTheme}>
			<Base {...props} />
		</ThemeProvider>
	);
};

export default Provider;
