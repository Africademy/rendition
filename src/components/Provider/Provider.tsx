import {
	Grommet
} from 'grommet'
import * as React from 'react';

const Provider = ({ theme, ...props }: any) => {
	return (
		<Grommet theme={theme} {...props} />
	);
};

export default Provider;
