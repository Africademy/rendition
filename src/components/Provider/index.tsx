import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';
import { BreakpointProvider } from './BreakpointProvider';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	const providerTheme = merge(defaultTheme, theme);
	return (
		<BreakpointProvider breakpoints={providerTheme.breakpoints}>
			<Base theme={providerTheme} {...props} />
		</BreakpointProvider>
	);
};

export interface ThemedProvider extends DefaultProps {
	theme?: Partial<Theme>;
}

export default Provider;
