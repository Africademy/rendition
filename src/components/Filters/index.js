import * as React from 'react'
import * as moment from 'moment'
import * as cloneDeep from 'lodash/cloneDeep'
import * as first from 'lodash/first'
import * as find from 'lodash/find'
import * as assign from 'lodash/assign'
import * as map from 'lodash/map'
import * as FaFilter from 'react-icons/lib/fa/filter'
import * as FaSearch from 'react-icons/lib/fa/search'
import styled from 'styled-components'
import FilterSummary from './Summary'
import ViewsMenu from './ViewsMenu'
import Button from '../Button'
import Modal from '../Modal'
import Select from '../Select'
import { Flex, Box } from '../Grid'
import * as utils from '../../utils'
import SchemaSieve from './SchemaSieve'
import PineTypes from '../PineTypes'

/**
 * The filter component requires the following props:
 * rules - an array of filter rule objects
 * schema - a SchemaSieve schema
 * views - an array of objects, each of which contains an array of predefined filter views,
 * setRules - a method that is called to set rules
 * setViews - a method that is called to set views
 */

const sieve = SchemaSieve()

const SimpleSearchBox = styled.div`
  position: relative;
  width: 500px;
  border-bottom: 2px solid #ccc;
  padding-left: 20px;
  padding-top: 3px;
  margin-left: 30px;
  margin-right: 30px;

  .search-icon {
    position: absolute;
    top: 7px;
    left: 0;
    color: #9b9b9b;
  }

  input {
    box-shadow: none;
    border: none;
    width: 100%;
    font-size: inherit;
    padding: 5px;
    ::placeholder {
      font-style: italic;
      color: #9b9b9b;
    }
  }
`

const FilterWrapper = styled(Box)`
  position: relative;
`

const FilterInput = props => {
  const PineTypeInput = PineTypes[props.type].Edit

  return (
    <PineTypeInput
      schema={props.schema}
      value={props.value}
      operator={props.operator}
      onChange={props.onChange}
    />
  )
}

class Filters extends React.Component {
  constructor (props) {
    super(props)
    this.handleEditChange = this.handleEditChange.bind(this)
    this.generateFreshEdit = this.generateFreshEdit.bind(this)

    const { rules } = this.props
    const existingRule = find(rules, { name: sieve.SIMPLE_SEARCH_NAME })

    this.state = {
      showModal: false,
      edit: this.generateFreshEdit(),
      searchString: (existingRule && existingRule.value) || ''
    }

    // Clean exsting rules on load
    this.filterAndSetRules(rules)
  }

  componentWillReceiveProps (nextProps) {
    const currentRules = nextProps.rules
    const existing = find(currentRules, { name: sieve.SIMPLE_SEARCH_NAME })
    if (existing) {
      const { value } = existing
      if (value !== this.state.searchString) {
        this.setState({ searchString: value })
      }
    } else {
      this.setState({ searchString: '' })
    }
  }

  /**
   * Remove filter rules that don't have a matching schema entry.
   * This can happen if the provided rules are from an old version of the
   * provided schema. Rather than making wrapper code handle this, we just clean
   * up rules as they are output from this component.
   */
  filterInvalidRules (rules) {
    return rules.filter(
      rule =>
        (rule.name === sieve.SIMPLE_SEARCH_NAME && rule.value) ||
        this.props.schema.hasOwnProperty(rule.name)
    )
  }

  filterAndSetRules (rules) {
    this.props.setRules(this.filterInvalidRules(rules))
  }

  generateFreshEdit () {
    if (!this.props.schema) {
      return {}
    }
    const inputModels = sieve.makeFilterInputs(this.props.schema)

    const edit = {
      name: Object.keys(inputModels).shift(),
      value: ''
    }

    edit.operator = inputModels[edit.name].availableOperators[0].value
    edit.label = inputModels[edit.name].label
    edit.type = inputModels[edit.name].type

    return this.setDefaultEditData(edit, edit.name)
  }

  addFilterRule (rule) {
    const { rules } = this.props
    rules.push(rule)
    this.filterAndSetRules(rules)
  }

  editFilterRule (rule) {
    const { rules } = this.props
    const updatedRules = rules.map(r => (r.id === rule.id ? rule : r))

    this.filterAndSetRules(updatedRules)
  }

  addRule (rule) {
    const inputModels = sieve.makeFilterInputs(this.props.schema)

    if (!rule) {
      rule = cloneDeep(this.state.edit)
    }
    const baseRule = inputModels[rule.name]
    const newRule = assign(cloneDeep(baseRule), rule)

    if (newRule.id) {
      this.editFilterRule(newRule)
    } else {
      newRule.id = utils.randomString()
      this.addFilterRule(newRule)
    }
    this.setState({
      showModal: false,
      edit: this.generateFreshEdit()
    })
  }

  updateSimpleSearch (val) {
    this.setState({ searchString: val })
    const { rules } = this.props
    const existingRule = find(rules, { name: sieve.SIMPLE_SEARCH_NAME })
    if (existingRule) {
      existingRule.value = val
      this.editFilterRule(existingRule)
    } else {
      this.addFilterRule({
        name: sieve.SIMPLE_SEARCH_NAME,
        value: val,
        id: utils.randomString()
      })
    }
  }

  showEditModal (rule) {
    this.setState({
      showModal: true,
      edit: cloneDeep(rule)
    })
  }

  removeRule (rule) {
    if (rule.name === sieve.SIMPLE_SEARCH_NAME) {
      this.setState({ searchString: '' })
    }

    const { rules } = this.props

    const updatedRules = rules.filter(r => r.id !== rule.id)
    this.filterAndSetRules(updatedRules)
  }

  setDefaultEditData (data, value) {
    const update = cloneDeep(data)
    const inputModels = sieve.makeFilterInputs(this.props.schema)
    const model = inputModels[value]
    update.type = model.type
    update.name = value
    update.operator = model.availableOperators[0].value
    update.label = model.label
    if (model.type === 'Date Time') {
      update.value = moment().format('YYYY-MM-DDTHH:mm')
    } else if (model.type === 'Date') {
      update.value = moment().format('YYYY-MM-DD')
    } else if (model.type === 'Time') {
      update.value = moment().format('HH:mm')
    } else if (model.type === 'Enum') {
      update.value = first(this.props.schema[model.name].values)
    } else {
      update.value = ''
    }

    return update
  }

  handleEditChange (value, attribute) {
    let update = this.state.edit

    if (attribute === 'name' && update.name !== value) {
      update = this.setDefaultEditData(update, value)
    } else if (attribute === 'operator') {
      update.value = null
      update[attribute] = value
    } else {
      update[attribute] = value
    }

    this.setState({ edit: update })
  }

  saveView (name, scopeKey) {
    const { rules } = this.props
    let { views } = this.props

    const newView = {
      name,
      rules: cloneDeep(rules),
      id: utils.randomString(),
      scopeKey
    }

    if (!views) {
      views = []
    }

    const store = views.find(item => item.key === scopeKey)

    store.data.push(newView)

    this.props.setViews(views)
  }

  deleteView (view, scopeKey) {
    const { views } = this.props

    const store = views.find(item => item.key === scopeKey)

    store.data = store.data.filter(v => v.id !== view.id)

    this.props.setViews(views)
  }

  showNewFilterModal () {
    this.setState({
      showModal: true,
      edit: this.generateFreshEdit()
    })
  }

  render () {
    const inputModels = sieve.makeFilterInputs(this.props.schema)
    const rules = this.props.rules || []

    return (
      <FilterWrapper mb={3}>
        <Flex justify='space-between'>
          <Button
            disabled={this.props.disabled}
            primary
            onClick={() => this.showNewFilterModal()}
            {...this.props.addFilterButtonProps}
          >
            <FaFilter style={{ marginRight: 10 }} />
            Add filter
          </Button>

          <SimpleSearchBox>
            <input
              disabled={this.props.disabled}
              placeholder='Search entries...'
              value={this.state.searchString}
              onChange={e => this.updateSimpleSearch(e.target.value)}
            />
            <FaSearch className='search-icon' name='search' />
          </SimpleSearchBox>

          <ViewsMenu
            buttonProps={this.props.viewsMenuButtonProps}
            disabled={this.props.disabled}
            rules={rules}
            views={this.props.views || []}
            schema={this.props.schema}
            setRules={rules => this.filterAndSetRules(rules)}
            deleteView={(view, scopeKey) => this.deleteView(view, scopeKey)}
          />
        </Flex>

        {this.state.showModal && (
          <div>
            <Modal
              title={
                this.state.edit.id
                  ? 'Update existing filter'
                  : 'Add a new filter'
              }
              cancel={() => this.setState({ showModal: false })}
              done={() => this.addRule()}
              action={this.state.edit.id ? 'Update filter' : 'Add filter'}
            >
              <form onSubmit={e => e.preventDefault() && this.addRule()}>
                <Flex>
                  <Select
                    mr={20}
                    value={this.state.edit.name}
                    onChange={e =>
                      this.handleEditChange(e.target.value, 'name')
                    }
                  >
                    {map(inputModels, ({ name, label }) => (
                      <option key={name} value={name}>
                        {label || name}
                      </option>
                    ))}
                  </Select>
                  <Select
                    mr={20}
                    value={this.state.edit.operator}
                    onChange={e =>
                      this.handleEditChange(e.target.value, 'operator')
                    }
                  >
                    {map(
                      inputModels[this.state.edit.name].availableOperators,
                      ({ value, label }) => (
                        <option value={value} key={value}>
                          {label}
                        </option>
                      )
                    )}
                  </Select>
                  {inputModels[this.state.edit.name].type !== 'Boolean' && (
                    <FilterInput
                      operator={this.state.edit.operator}
                      schema={this.props.schema[this.state.edit.name]}
                      value={this.state.edit.value}
                      onChange={value => this.handleEditChange(value, 'value')}
                      type={inputModels[this.state.edit.name].type}
                    />
                  )}
                </Flex>
              </form>
            </Modal>
          </div>
        )}

        {!!rules.length &&
          !this.props.disabled && (
            <FilterSummary
              edit={rule => this.showEditModal(rule)}
              delete={rule => this.removeRule(rule)}
              saveView={(name, scope) => this.saveView(name, scope)}
              rules={rules}
              views={this.props.views || []}
              schema={this.props.schema}
            />
          )}
      </FilterWrapper>
    )
  }
}

export default Filters
