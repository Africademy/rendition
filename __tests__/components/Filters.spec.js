/* globals expect, describe, it */
import { mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Filters from '../../src/components/Filters'
import FiltersSummary from '../../src/components/Filters/Summary'
import * as SchemaSieve from '../../src/components/Filters/SchemaSieve'
import { ViewListItem } from '../../src/components/Filters/ViewsMenu'

const schema = {
  type: 'object',
  properties: {
    Name: {
      title: 'Pokemon Name',
      type: 'string'
    }
  }
}

const filter = SchemaSieve.createFilter(schema, [{
  field: 'Name',
  operator: 'is',
  value: 'Squirtle'
}])

const view = {
  id: 'abcde',
  name: 'Test',
  scope: null,
  filters: [ filter ]
}

describe('Filters component', () => {
  it('should match the stored snapshot', () => {
    const tree = renderer.create(
      <Provider>
        <Filters schema={schema} />
      </Provider>
    )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('filters property', () => {
    it('should not render a summary if there are no filters', () => {
      const component = mount(
        <Provider>
          <Filters
            schema={schema}
          />
        </Provider>
      )

      expect(component.find(FiltersSummary)).toHaveLength(0)

      component.unmount()
    })

    it('should render a summary of filters that are provided as props', () => {
      const component = mount(
        <Provider>
          <Filters
            schema={schema}
            filters={[filter]}
          />
        </Provider>
      )

      expect(component.find(FiltersSummary)).toHaveLength(1)

      component.unmount()
    })

    it('should render a summary of filters that are added as props after the component mounts', () => {
      const component = mount(
        <Filters
          schema={schema}
        />
      )

      component.setProps({ filters: [ filter ] })

      component.update()

      expect(component.find(FiltersSummary)).toHaveLength(1)

      component.unmount()
    })

    it('should not render a summary of filters if they are removed after the component mounts', () => {
      const component = mount(
        <Filters
          schema={schema}
          filters={[filter]}
        />
      )

      component.setProps({ filters: null })

      component.update()

      expect(component.find(FiltersSummary)).toHaveLength(0)
      component.unmount()
    })
  })

  describe('views property', () => {
    it('should not render a views menu if there are no views', () => {
      const component = mount(
        <Provider>
          <Filters
            schema={schema}
          />
        </Provider>
      )

      component.find('button').at(1).simulate('click')
      expect(component.find(ViewListItem)).toHaveLength(0)
      component.unmount()
    })

    it('should render a list of views that are provided as props', () => {
      const component = mount(
        <Provider>
          <Filters
            schema={schema}
            views={[view]}
          />
        </Provider>
      )

      component.find('button').at(1).simulate('click')
      expect(component.find(ViewListItem)).toHaveLength(1)
      component.unmount()
    })

    it('should render a list of views that are added as props after the component mounts', () => {
      const component = mount(
        <Filters
          schema={schema}
        />
      )

      component.setProps({ views: [ view ] })
      component.update()
      component.find('button').at(1).simulate('click')
      expect(component.find(ViewListItem)).toHaveLength(1)
      component.unmount()
    })

    it('should not render a list of views if they are removed after the component mounts', () => {
      const component = mount(
        <Filters
          schema={schema}
          views={[view]}
        />
      )

      component.setProps({ views: null })
      component.update()
      component.find('button').at(1).simulate('click')
      expect(component.find(ViewListItem)).toHaveLength(0)
      component.unmount()
    })

    it('should render when the schema contains an unknown type', () => {
      const unknownSchema = {
        type: 'object',
        properties: {
          test: {
            type: 'Foo Bar'
          }
        }
      }
      const component = mount(
        <Filters
          schema={unknownSchema}
        />
      )

      component.unmount()
    })
  })
})
