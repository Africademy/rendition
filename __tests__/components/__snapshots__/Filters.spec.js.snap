// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component should match the stored snapshot 1`] = `
.c4 {
  box-sizing: border-box;
}

.c3 {
  margin-bottom: 16px;
}

.c18 {
  margin-top: 1px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  outline: none;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 1px solid #00AEEF;
  border-radius: 20px;
  color: #444444;
  padding: 4px 30px;
  font-size: 14px;
  line-height: 1.5;
  background: #00AEEF;
  color: #444444;
  border-radius: 20px;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c8:hover {
  box-shadow: 0px 0px 0px 2px #00AEEF;
}

.c16 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  outline: none;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 1px solid #00AEEF;
  border-radius: 20px;
  color: #444444;
  padding: 4px 30px;
  font-size: 14px;
  line-height: 1.5;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c16:hover {
  box-shadow: 0px 0px 0px 2px #00AEEF;
}

.c1 {
  font-family: 'Nunito',Arial,sans-serif;
  font-size: 14px;
  line-height: 1.5;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c7 {
  font-family: CircularStd,Arial,sans-serif;
  font-weight: 500;
  height: 38px;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: white;
}

.c7 svg {
  color: inherit !important;
  font-size: 0.875em;
}

.c7:disabled {
  cursor: not-allowed;
}

.c7:hover:enabled,
.c7:focus:enabled,
.c7:active:enabled {
  box-shadow: none;
  background: hsl(196.29999999999995,100%,37.5%);
  border-color: hsl(196.29999999999995,100%,37.5%);
  color: white;
  opacity: undefined;
}

.c15 {
  font-family: CircularStd,Arial,sans-serif;
  font-weight: 500;
  height: 38px;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #00AEEF;
  border-color: #00AEEF;
}

.c15 svg {
  color: inherit !important;
  font-size: 0.875em;
}

.c15:disabled {
  cursor: not-allowed;
}

.c15:hover:enabled,
.c15:focus:enabled,
.c15:active:enabled {
  box-shadow: none;
  background: #00AEEF;
  border-color: #00AEEF;
  color: white;
  opacity: undefined;
}

.c6 {
  margin-right: 30px;
}

.c14 {
  padding-left: 16px;
  padding-right: 0;
}

.c10 {
  position: relative;
  width: 100%;
  border-bottom: 1px solid #c6c8c9;
  padding-left: 24px;
  padding-top: 3px;
}

.c10 .search-icon {
  color: #c6c8c9;
  font-size: 0.9em;
  position: absolute;
  top: 50%;
  left: 4px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c10 input {
  outline: none;
  background: transparent;
  box-shadow: none;
  border: none;
  width: 100%;
  font-size: inherit;
  padding: 5px 5px 8px;
  height: auto;
}

.c10 input:hover {
  box-shadow: none;
}

.c10 input::-webkit-input-placeholder {
  color: #c6c8c9;
  font-weight: 300;
}

.c10 input::-moz-placeholder {
  color: #c6c8c9;
  font-weight: 300;
}

.c10 input:-ms-input-placeholder {
  color: #c6c8c9;
  font-weight: 300;
}

.c10 input::placeholder {
  color: #c6c8c9;
  font-weight: 300;
}

.c12 {
  display: inline-block;
  border-radius: 3px;
  vertical-align: top;
  position: relative;
}

.c19 {
  width: 28px;
}

.c13 {
  margin: 0;
  width: 100%;
}

.c11 {
  margin-left: 30px;
}

.c9 {
  -webkit-flex-basis: 500px;
  -ms-flex-preferred-size: 500px;
  flex-basis: 500px;
}

.c2 {
  position: relative;
}

.c0 {
  font-family: 'Nunito',Arial,sans-serif;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4,
.c0 h5,
.c0 h6,
.c0 label,
.c0 button {
  font-family: CircularStd,Arial,sans-serif;
}

<div
  className="c0 c1"
>
  <div
    className="c2 c3 c4"
  >
    <div
      className="c5 c4"
    >
      <button
        className="c6 c7 c8"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "marginRight": 10,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m35.8 6.6q0.4 0.9-0.3 1.5l-11 11.1v16.5q0 1-0.9 1.3-0.3 0.1-0.5 0.1-0.6 0-1-0.4l-5.7-5.7q-0.5-0.4-0.5-1v-10.8l-11-11q-0.7-0.7-0.3-1.6 0.4-0.9 1.3-0.9h28.6q0.9 0 1.3 0.9z"
            />
          </g>
        </svg>
        Add filter
      </button>
      <div
        className="c9"
      >
        <div
          className="c10"
        >
          <input
            onChange={[Function]}
            placeholder="Search entries..."
            style={
              Object {
                "color": undefined,
              }
            }
            value=""
          />
          <svg
            className="search-icon"
            fill="currentColor"
            height="1em"
            name="search"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m27.2 18.6q0-4.2-2.9-7.1t-7.1-2.9-7 2.9-3 7.1 2.9 7 7.1 3 7.1-3 2.9-7z m11.4 18.5q0 1.2-0.8 2.1t-2 0.8q-1.2 0-2-0.8l-7.7-7.7q-4 2.8-8.9 2.8-3.2 0-6.1-1.3t-5-3.3-3.4-5-1.2-6.1 1.2-6.1 3.4-5.1 5-3.3 6.1-1.2 6.1 1.2 5 3.3 3.4 5.1 1.2 6.1q0 4.9-2.7 8.9l7.6 7.6q0.8 0.9 0.8 2z"
              />
            </g>
          </svg>
        </div>
      </div>
      <div
        className=""
      >
        <div
          className="c11 c12 c4"
        >
          <button
            className="c13 c14 c15 c16"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            type="button"
          >
            <div
              className="c17 c4"
            >
              <div
                className="c18 c4"
              >
                <span>
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "marginRight": 10,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m17.6 19.9l12.2 12.2q-2.3 2.4-5.5 3.7t-6.7 1.3q-4.6 0-8.6-2.3t-6.2-6.2-2.3-8.6 2.3-8.6 6.2-6.2 8.6-2.3v17z m4.2 0.1h17.3q0 3.5-1.4 6.7t-3.7 5.5z m15.8-2.9h-17.1v-17.1q4.7 0 8.6 2.3t6.2 6.2 2.3 8.6z"
                      />
                    </g>
                  </svg>
                  Views
                </span>
              </div>
              <span
                className="c19"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "color": undefined,
                      "verticalAlign": "middle",
                    }
                  }
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m31.4 15.7q0 0.6-0.5 1l-10 10q-0.4 0.4-1 0.4t-1-0.4l-10-10q-0.4-0.4-0.4-1t0.4-1 1-0.4h20q0.6 0 1 0.4t0.5 1z"
                    />
                  </g>
                </svg>
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
